File: CPP/No437.cpp
/**
 * Created by Xiaozhong on 2020/9/20.
 * Copyright (c) 2020/9/20 Xiaozhong. All rights reserved.
 */

#include "iostream"
#include "vector"
#include "algorithm"
#include "unordered_map"

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    int count = 0;

    void dfs(TreeNode *root, int target, int total, unordered_map<int, int> &mapper) {
        int sum = root->val + total;

        // 本身节点已经达到了 target
        if (sum == target) ++count;
        // 在当前节点查看前缀和
        if (mapper[sum - target] > 0) count += mapper[sum - target];
        if (root->left) {
            mapper[sum] = mapper[sum] + 1;
            dfs(root->left, target, sum, mapper);
            mapper[sum] = mapper[sum] - 1;  // 回溯
        }
        if (root->right) {
            mapper[sum] = mapper[sum] + 1;
            dfs(root->right, target, sum ,mapper);
            mapper[sum] = mapper[sum] - 1;  // 回溯
        }

    }

public:
    int pathSum(TreeNode *root, int sum) {
        if (!root) return 0;
        unordered_map<int, int> mapper;
        dfs(root, sum, 0, mapper);
        return count;
    }
};

File: Java/src/No437.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.HashMap;
import java.util.Map;

public class No437 {
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private Map<Integer, Integer> mapper = new HashMap<>();
    private int count = 0;
    private int targetSum = 0;

    public int pathSum(TreeNode root, int targetSum) {
        if (root == null) return 0;
        this.targetSum = targetSum;
        dfs(root, 0);
        return count;
    }

    private void dfs(TreeNode root, int curr) {
        int sum = curr + root.val;
        if (sum == targetSum) count++;
        // 下面这句有疑问
        // 现在没了，因为是前缀和，需要在前面找到 sum - prev = target 的 prev
        if (mapper.containsKey(sum - targetSum)) count += mapper.get(sum - targetSum);
        if (root.left != null) {
            mapper.put(sum, mapper.getOrDefault(sum, 0) + 1);
            dfs(root.left, sum);
            mapper.put(sum, mapper.getOrDefault(sum, 0) - 1);
        }
        if (root.right != null) {
            mapper.put(sum, mapper.getOrDefault(sum, 0) + 1);
            dfs(root.right, sum);
            mapper.put(sum, mapper.getOrDefault(sum, 0) - 1);
        }
    }
}
